### ISSUE TITLE
Potential Example Application:  Predicting GPS Satellite Ephemeris

### ISSUE THREAD

**Issue by musicarroll** [dynamicslab/pysindy#167]
## Is your feature request related to a problem? Please describe.

Potential Example Application:  Predicting GPS Satellite Ephemeris.  I've created a script that uses pysindy to train a model on NGA precise ephemeris data.  See https://github.com/musicarroll/gps/blob/main/pspredict.py
## Describe the solution you'd like

Would love some help fine tuning of the model, leveraging Alan's awesome knowledge!  The preliminary model derived uses polynomial and fourier libraries.  This prelim model achieves a mean squared error of about 55 m, whereas, NGA's own 9-day predict seems to be at least half of that.  A custom library may be in order along with some fine tuning of the optimizers.  I also plan to try constrained optimization constraining the derivatives of position data to be the NGA-supplied velocity data.
Challenge:  Show that a pysindy model can outperform NGA's 9-day predictions!

## Describe alternatives you've considered

I've tried using the identify library, the fourier library alone, and tensoring polynomial and fourier libraries, all with markedly worse performance.

## Additional context

See https://github.com/musicarroll/gps/blob/main/plots/truth%20vs%20predict%20normed%20pos%20vector%202022-03-02%20135643.png

**Comment by akaptano** [comment]
Very cool problem -- I'll take a look at this over the weekend. One thing that it looks like you need to deal with, is that, from your "additional context", the position vector is basically a large sine wave with amplitude in the 100's of km, but the errors are order ~ 10 meters. In other words, you are trying to improve your prediction of the dynamics occurring on much smaller scales than the primary dynamics, which are occurring on the scale of 100's of km. Two things immediately to try:
1. Using the L1 or L2 norm in the optimizer, since using the L0 norm may truncate off those small-scale dynamics you want. 
2. Using normalize_columns=True in the optimizer, since this helps equalize the scales (I see you are doing this with STLSQ, but not with the other commented optimizers).

**Comment by musicarroll** [comment]
He thank, Alan!  I appreciate the prompt response.  I will try those things you suggest.
Also, I didn't mention that the model.simulate() call takes forever and then craps out.  I'm wondering if scaling and integrator step sizing might be the culprit. 
Thanks again and have great weekend.

**Comment by musicarroll** [comment]
Got some small improvement with
    optimizer = ps.SR3(threshold=0.01,thresholder="l2", normalize_columns=True, tol=1e-9)
Results:

Space Vehicle:  PRN  10
Size of training set: 30.0 days.
Size of test set: 9.0 days.


Model (derived using pysindy):
(x)' = -0.183 1 + -0.002 sin(1 x) + 0.002 cos(1 x) + 0.002 sin(1 y) + -0.002 cos(1 y) + 0.004 sin(1 z) + 0.005 sin(2 x) + -0.002 cos(2 x) + 0.006 sin(2 y) + 0.003 cos(2 y) + 0.002 sin(2 z) + -0.003 cos(2 z)
(y)' = 0.008 1 + 0.002 sin(1 x) + -0.001 cos(1 x) + -0.003 sin(1 y) + -0.002 cos(1 y) + 0.001 sin(1 z) + -0.005 cos(1 z) + -0.002 sin(2 x) + -0.008 cos(2 x) + 0.006 sin(2 y) + -0.004 cos(2 y) + -0.001 sin(2 z) + 0.002 cos(2 z)
(z)' = -252.792 1 + 0.001 sin(1 x) + 0.001 sin(1 y) + 0.001 cos(2 x)

Model Score (MSE): 0.0493 km

Model Score (MSE): 49.2927 m

Model Score (MSE): 4929.2746 cm

Maybe threshold should be smaller?  Will try.

**Comment by musicarroll** [comment]
Subtract off the mean of the velocity field from each velocity item?  Mean across the constellation or mean across the whole training period for each given satellite?  For computing a metric called user range error (URE) it is common to subtract off the constellation mean of the radial error minus clock error (in distance terms) at each 5min epoch from the raw radial error minus clock error for each satellite.

**Comment by akaptano** [comment]
I think you know better than I do what makes sense for a mean-subtraction of the data. The point here is choosing some way to kill off the relatively tame variation over the 100 of kms scale, and just build a model for the small scales. But maybe this won't work well.

**Comment by musicarroll** [comment]
Copy that.  I'll try some things.  Also wondering if I put in a another variable and data expressing time-dependent phase shifting (or orbital drift) will pysindy correctly insert the phase shift inside of the sine and cosines.  I'm also going to experiment with adding the clock bias variable (it's on the order of microsec, but can have a major impact on the underlying pseudorange measurements.  The data I'm using here are not raw measurement data;  they are estimates from NGA.  But I'm treating them as if they were raw measurements because the bottom line is to just be able to predict them -- whatever they are.  Thanks for your suggestions!  Very much appreciated!

**Comment by musicarroll** [comment]
Created new script that uses position and clock terms (and clock rate with the derivatives):
https://github.com/musicarroll/gps/blob/main/pspredict_clock.py
Actually shaves off almost 10m from the MSE.
Now I'll start playing with means (and maybe add some derived phase shift data).

**Comment by akaptano** [comment]
Looking forward to seeing more, but closing this for now since I'm not sure this qualifies as "general interest" for the community, and it is not an active discussion. Happy to discuss more in the closed issue though.

**Comment by musicarroll** [comment]
No problem.  Appreciate your help.  I got sidetracked but will return to it
in the next couple of weeks.
